### ISSUE TITLE
Usage with Nix Flakes

### ISSUE THREAD

**Issue by shadowrylander** [DavHau/mach-nix#153]
Hello!

I was just wondering how I would go about using this with Nix Flakes, as in what module I would have to import in `nixpkgs.lib.nixosSystem.modules`? Would `mach-nix.nixosModules.mach-nix` work?

Thank you kindly for the help!

**Comment by DavHau** [comment]
Mach-nix does currently not propose a module. Also I'm not sure what would be the benefit. Mach-nix is just a package. There is nothing that needs to be integrated with the system.

**Comment by shadowrylander** [comment]
Ah; could you show me how to use it in a flake for the NixOS Configuration, then? I'm a bit confused on how to pass it to the config; do I just add it to the list of arguments I pass to `nixpkgs.lib.nixosSystem`?

**Comment by DavHau** [comment]
A nixos configuration with mach-nix inside the system environment could look like this:
```nix
{ pkgs, ... }:
let
  mach-nix =  (import (builtins.fetchGit {
    url = "https://github.com/DavHau/mach-nix.git";
    ref = "refs/heads/master" ;
    rev ="70b6328ca28b54a135a19d8d7267a3d1639c3b3d";
  })).mach-nix;
in
{
  environment.systemPackages = [
    mach-nix
  ];
}
```

**Comment by shadowrylander** [comment]
But what if I were to include it in the inputs set in `flake.nix`?  How would the respective output set look?

**Comment by DavHau** [comment]
Maybe it helps to play around with it in `nix repl` a bit:
```
nix-repl> f = (builtins.getFlake "github:davhau/mach-nix")     

nix-repl> f.packages.x86_64-linux.mach-nix
«derivation /nix/store/a9wgylch2j0j59bf7blbcnm4323zi86f-python3.7-mach-nix-2.4.1.drv»


```

**Comment by shadowrylander** [comment]
Yeah, I'm still kinda' stuck on where to put `[f].packages.x86_64-linux.mach-nix`...

**Comment by DavHau** [comment]
Sorry, I was busy finishing the 3.0.0 release. Were you able to solve your problem in the meantime?

**Comment by shadowrylander** [comment]
I don't want to keep manually updating the module rev and hash, though... It's why I use the master branch everywhere; I wanted to copy a rolling-release model. Flakes seemed to finally be the easiest way to do it.

**Comment by shadowrylander** [comment]
Actually, I figured I'll just pass mach-nix directly to my overlays as an argument, so problem solved! More or less...

**Comment by shadowrylander** [comment]
Turns out I get stuck on `attribute 'mkPython' missing` when using the current flake and passing `mach-nix.packages.${system}.mach-nix`.

**Comment by shadowrylander** [comment]
Pinging @DavHau! Sorry, I'm really excited for this! 😅😻

**Comment by DavHau** [comment]
I think the problem is that the mach-nix flake doesn't expose the `mkPython` function.
You can add it here and make a PR if you like
(https://github.com/DavHau/mach-nix/blob/master/flake.nix#L19-L25)

**Comment by shadowrylander** [comment]
How would I do that, though? If I'm passing `mach-nix.packages.mach-nix` to my module, and my module calls `mach-nix.packages.mach-nix.mkPython`, i.e. *that* module's `mkPython`?

**Comment by newkozlukov** [comment]
> I think the problem is that the mach-nix flake doesn't expose the `mkPython` function.
> You can add it here and make a PR if you like
> (https://github.com/DavHau/mach-nix/blob/master/flake.nix#L19-L25)

Hi! I've run into this too. I was thinking about exposing `mkPython` through the `lib` flake output, but so far didn't succeed (hadn't much time to investigate this yet): https://github.com/newkozlukov/mach-nix/pull/1/

To keep the graph strongly connected, I also raised the question in https://discourse.nixos.org/t/flake-nix-interface-nix-show-fails-with-aarch64-linux-required-to-build/9561 but this issue seems much more appropriate place for discussion

Thanks!

**Comment by shadowrylander** [comment]
The way I'm thinking of doing it is either by passing my module all the `mk` functions individually, or creating a new set with all the `mk` functions included as attributes.

**Comment by DavHau** [comment]
@newkozlukov Thanks for looking into this! Feel free to just open a WIP PR with your changes and we can try to fix it from there.

**Comment by shadowrylander** [comment]
It seems as though none of my changes work; any help on them would be greatly appreciated!

**Comment by newkozlukov** [comment]
@shadowrylander Tbh, I'm pretty occupied lately and made no progress since last comment

**Comment by shadowrylander** [comment]
Aw! 😹 Ah well; whenever you get around to it, if you do!

**Comment by nat543207** [comment]
Flake inputs actually have some special handling that causes `toString input` to return a Nix store path; as a result, you can import Nix files from flake inputs directly.  I've been using that strategy in my own flake-based projects; a working (but very simple) example flake that does this with `mach-nix` is shown below:
```
{
  description = "Using mach-nix from a flake";

  inputs = {
    nixpkgs.url = github:NixOS/nixpkgs/nixos-unstable;
    mach-nix.url = github:DavHau/mach-nix/f80ee22; # v3.0.0 release tag
  };

  outputs = { self, nixpkgs, mach-nix }: let
    pkgs = (import nixpkgs { system = "x86_64-linux"; }).pkgs;
    mach-nix-utils = import mach-nix {
      inherit pkgs;
      python = "python3";
    };
  in {

    packages.x86_64-linux.customPython = mach-nix-utils.mkPython {
      requirements = ''
        pytest
      '';
    };

    defaultPackage.x86_64-linux = self.packages.x86_64-linux.customPython;
  };
}
```

`mkPython` works just fine, which you can see for yourself by initializing a repo with the above flake and running
```
[nat543207@hostname|tmp] nix shell .#customPython                                                                                                                                                             
[nat543207@hostname|tmp] python
Python 3.8.6 (default, Sep 23 2020, 13:54:27)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pytest
>>> pytest.fail('Intentional failure')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/6qbpa7k8mlgzb8gf67lms6ypx4q6vfp2-python3-3.8.6-env/lib/python3.8/site-packages/_pytest/outcomes.py", line 153, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: Intentional failure
>>>
```

However, functions like `buildPythonApplication` that rely on `extractor` don't run correctly due to a [`<nixpkgs>` import](https://github.com/DavHau/mach-nix/blob/9aa62d0d03587a2fb0accae70e4c036df51b0e13/lib/extractor/default.nix#L2) that can't be evaluated in pure-mode by the flake.  I've got a local patch to replace that import with the `pkgs` value used to initialize `mach-nix`; it hasn't caused me any issues yet, so I'll see if I can put a PR up for it tomorrow

**Comment by shadowrylander** [comment]
If I may ask something slightly off-topic: if anyone here has experience extending the default `lib` module, could you help me over on [reddit](https://www.reddit.com/r/NixOS/comments/jjhw44/review_lib_module_extension/?utm_medium=android_app&utm_source=share) with a major roadblock I'm having? It seems nobody else has the necessary time or experience to help me there on this particular matter.

Afterwards, I can continue helping debug the `mach-nix` flake, as the aforementioned issue seems to be stopping me from doing so.

**Comment by shadowrylander** [comment]
I can confirm that @nat543207's method works.

**Comment by shadowrylander** [comment]
I suppose I'll close this then!

**Comment by DavHau** [comment]
I merged some commits of @nat543207 and @newkozlukov and added the library functions to flakes.nix.

**Comment by ae-mo** [comment]
Hey, I've tried the approach @nat543207 suggested, but `nix develop` fails with the following error output:

```
building '/nix/store/gpc8k09imm8yki7xzi9kx8c2c3m5fbi4-python3.8-scs-2.1.2.drv'...
error: --- Error --- nix-daemon
builder for '/nix/store/gpc8k09imm8yki7xzi9kx8c2c3m5fbi4-python3.8-scs-2.1.2.drv' failed with exit code 1; last 10 log lines:
  ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-libffi-3.3/lib'
  ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-gdbm-1.18.1/lib'
  ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-sqlite-3.33.0/lib'
  ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-readline-6.3p08/lib'
  ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-ncurses-6.2/lib'
  ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-openssl-1.1.1h/lib'
  ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-configd-osx-10.8.5/lib'
  ld: framework not found Accelerate
  clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
  error: Command "clang -bundle -undefined dynamic_lookup -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-zlib-1.2.11/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-bzip2-1.0.6.0.1/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-expat-2.2.10/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-xz-5.2.5/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-libffi-3.3/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-gdbm-1.18.1/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-sqlite-3.33.0/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-readline-6.3p08/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-ncurses-6.2/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-openssl-1.1.1h/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-configd-osx-10.8.5/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-zlib-1.2.11/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-bzip2-1.0.6.0.1/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-expat-2.2.10/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-xz-5.2.5/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-libffi-3.3/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-gdbm-1.18.1/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-sqlite-3.33.0/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-readline-6.3p08/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-ncurses-6.2/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-openssl-1.1.1h/lib -L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-configd-osx-10.8.5/lib build/temp.macosx-10.6-x86_64-3.8/src/scsmodule.o build/temp.macosx-10.6-x86_64-3.8/scs/src/util.o build/temp.macosx-10.6-x86_64-3.8/scs/src/linalg.o build/temp.macosx-10.6-x86_64-3.8/scs/src/scs_version.o build/temp.macosx-10.6-x86_64-3.8/scs/src/cones.o build/temp.macosx-10.6-x86_64-3.8/scs/src/ctrlc.o build/temp.macosx-10.6-x86_64-3.8/scs/src/aa.o build/temp.macosx-10.6-x86_64-3.8/scs/src/normalize.o build/temp.macosx-10.6-x86_64-3.8/scs/src/scs.o build/temp.macosx-10.6-x86_64-3.8/scs/src/rw.o build/temp.macosx-10.6-x86_64-3.8/scs/linsys/common.o -L/nix/store/0mabx9absir73mmijrsv0dsyvyya463r-python3-3.8.6/lib -o build/lib.macosx-10.6-x86_64-3.8/_scs_direct.cpython-38-darwin.so -Wl,-framework -Wl,Accelerate -Wl,-framework -Wl,Accelerate" failed with exit status 1
error: --- Error --- nix-daemon
1 dependencies of derivation '/nix/store/xzj8zdqali8vk9xb6mzic3igrw57awwj-python3-3.8.6-env.drv' failed to build
error: --- Error -------------------------------------------------------------------------------------------------- nix
1 dependencies of derivation '/nix/store/brszc18sx6k91x4l1dcxr87i8nmzb0cq-mach-nix-python-shell-env.drv' failed to build
```

My `flake.nix`:
```
{
  description = "Flake for portmanteau - portfolio manager to end all uncertainties";

  inputs = {
    flake-utils.url = "github:numtide/flake-utils";
    mach-nix = {
      url = github:DavHau/mach-nix/1ec92303acd142aa1a3b60bb97745544cf049312; # v3.1.1 release tag
      inputs.nixpkgs.follows = "nixpkgs";
      inputs.flake-utils.follows = "flake-utils";
    };
  };
  outputs = { self, nixpkgs, mach-nix, flake-utils }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = (import nixpkgs { inherit system; }).pkgs;
        mach-nix-utils = import mach-nix {
          inherit pkgs;
          python = "python3";
        };
        requirements = ''
            pyportfolioopt
            pytest
          '';
      in rec
      {

        devShell = mach-nix-utils.mkPythonShell {
          inherit requirements;
        };

        packages.portmanteau = mach-nix-utils.mkPython {
          inherit requirements;
        };

        defaultPackage = packages.portmanteau;

      }
  );
}
```

I'm on macOS. Any ideas?

**Comment by shadowrylander** [comment]
@ae-mo You might want to check on that error; it seems to be screaming. 😹

**Comment by ae-mo** [comment]
It seems to be a problem related to the `scs` python package, not to the flake itself, so I guess this is not the place to discuss it 😄

**Comment by DavHau** [comment]
@ae-mo you could try settings the provider for scs to nixpkgs. Or try with the 'conda-beta' brach of mach-nix which can take packages from conda.

**Comment by ae-mo** [comment]
@DavHau I switched to Linux in the end and it worked, after fiddling with providers. Thanks!

**Comment by SomeoneSerge** [comment]
Run into this again and it wasn't immediately obvious to me how to use this, so posting the snippet here (as probably the easiest googlable place).

TLDR is `flake-utils.eachSystem` transforms `lib` into `outputs.lib.${system}`, so one should currently use e.g. `mach-nix.lib.${system}.mkPython`.

Note that `poetry2nix` takes a different approach and puts their convenience functions into overlays instead. This possibly makes sense, as `lib` is (probably; better ask Nix maintainers) supposed to be a place for `system`-independent outputs:

- `poetry2nix`:
  - [flake.nix](https://github.com/nix-community/poetry2nix/blob/01c042021c4a22c5ba6537adf0aed8f689c95ae3/flake.nix#L14)
  - the output that has `mk*` functions: [overlay.nix](https://github.com/nix-community/poetry2nix/blob/01c042021c4a22c5ba6537adf0aed8f689c95ae3/overlay.nix#L3), accessible via `pkgs` in your nix files
- [`flake-utils`](https://github.com/numtide/flake-utils/blob/3982c9903e93927c2164caa727cd3f6a0e6d14cc/flake.nix#L4) just puts a flat top-level `lib`

* * *
## Snippet that currently works

Step 0:

```
# flake.nix
{
  description = "...";
  
  inputs.flake-utils.url = "github:numtide/flake-utils";
  inputs.mach-nix.url = "github:DavHau/mach-nix";
  inputs.mach-nix.inputs.nixpkgs.follows = "nixpkgs";

  outputs = { self, nixpkgs, flake-utils, mach-nix }: (flake-utils.lib.eachDefaultSystem (system:
  let
    pkgs = import nixpkgs { inherit system; };
  in {
    devShell = (import ./shell.nix { nixpkgs=pkgs; mach-nix=mach-nix.lib.${system}; });
  }));
}
```

Step 1:

```
# shell.nix

{ mach-nix, nixpkgs ? (import <nixpkgs> { })}:

with nixpkgs;

let
  python = python3;
  myEnv = mach-nix.mkPythonShell {
    requirements = ''
      opencv-python
      numpy
      msgpack
      asyncio-nats-client
      '';
  };
in myEnv
```

Step 2:
Happily run `nix develop`

**Comment by DavHau** [comment]
@SomeoneSerge Thanks for noticing this.
Currently the `lib` functions of mach-nix are in fact system dependent. They all return a derivation and that derivation is built via a certain nixpkgs.
We could remove lib's dependency on nixpkgs, and instead make nixpkgs a mandatory argument whenever you call one of the lib functions. But then you would ignore the nixpkgs specified via your flakes.lock. Would you really want that?

**Comment by SomeoneSerge** [comment]
@DavHau  Hmm, if `nixpkgs` is made an argument of `mach-nix.lib.mk*`, then it seems they all would get the client flake's `nixpkgs` input (pinned in user's `flake.lock`) opposed to `inputs.mach-nix.inputs.nixpkgs` (pinned in `mach-nix`'s `flake.lock`) - doesn't look criminal to me.

Does anything in `mk*` assume specific nixpkgs revisions, like attribute names? What we could lose is that right now mach-nix's `flake.lock` provides a combination of `(mach-nix revision, nixpkgs revision)` that is guaranteed/expected to work, whereas taking `nixpkgs` from the user means that when the user runs her build the first time (before her `flake.lock` is generated) it may possibly fail
